Documentation/General/emergentArchitectureDesign/emergentArchitectureDesign.tex

```diff
@@ -0,0 +1,180 @@
+The backend keeps track of which rooms are active and what song is currently played in the room.
+Furthermore, the backend is responsible for transmitting the chat messages between the users in the room.
+An example interaction between the frontend and the backend is given in \ref{fig: backendInteraction}.
+% I am not so happy that we do not mention how we rely on Soundcloud for the streaming here, but I have the feeling it does not really fit here. Suggestions?
+
+\begin{figure}[h!]
+\includegraphics[scale=.4]{backendInteraction.png}
+\caption{Example interaction frontend and backend}
+\label{fig: backendInteraction}
+\end{figure}
+
+
+\par
+The backend and the frontend communicate over \Gls{HTTP} requests in the \Gls{JSON} format.
+% We might want to note Websockets at a later point
+The REST API will be written in Java using JAX-RS and Jackson.
+Jackson serializes Java objects to a JSON string based on their attributes, and vice versa: it is able to deserialize a Java object from a JSON string.
+This allows us to use JSON in the frondend, which is ideal in \Gls{JS}, but use classes and objects in Java, which greatly improves the testability of the responses.
+The backend will be tested using Unit tests written using JUnit and Mockito.
+
+\subsubsection{Music matching algorithm}\label{MatchingAlgorithm}
+The Music matching algorithm determines and compares moods for songs, rooms and users.
+Moodcat will be developed around a learning algorithm.
+This means that Moodcat will improve over time.
+Default values will be determined by examining the audio features provided by our data set.
+Then, through a system of user votes, the system will learn and improve over time.
+
+
+\section{Hardware/software mapping}
+For navigating through the user interface and posting messages in the chatroom a keyboard and mouse, or a touch screen with onscreen keyboard is required.
+We will annotate our elements with proper \Gls{ARIA} accessibilty attributes so that assistive technologies, such as screen readers, will work with our service properly.
+
+\par
+Moodcat requires an audio interface to play music.
+To interact with the audio interface from the browser, the Web Audio API\cite{WebAudioAPI} and NgAudio are used.
+The users browser should support these techniques in order to use Moodcat.
+
+\section{Persistent data management}
+Moodcat has to handle with both song data and chatroom data.
+
+\par
+For the song we store the basic properties, like artist, song title, genre and duration.
+Furthermore each song has its own valence and arousal value, which is computed and corrected by the music matching algorithm.
+The algorithm uses some of the audio features of the track for the computation of the valence and arousal values, these features are stored with the song.
+
+\par
+For the chatrooms we store basic properties, the currently playing song, the room members and a chat message.
```

Comment:
In my opinion, there is a semantic difference. the chat does not play the music, so in my head that makes a room contain a song and a chat(room). However, reading this again I think the message in this case is clear enough